server/src/main/java/org/elasticsearch/action/admin/cluster/stats/CCSUsageTelemetry.java

```diff
@@ -0,0 +1,246 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.action.admin.cluster.stats;
+
+import org.elasticsearch.common.util.Maps;
+
+import java.util.Collections;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.LongAdder;
+
+/**
+ * Service holding accumulated CCS search usage statistics. Individual cross-cluster searches will pass
+ * CCSUsage data here to have it collated and aggregated. Snapshots of the current CCS Telemetry Usage
+ * can be obtained by getting {@link CCSTelemetrySnapshot} objects.
+ * <br>
+ * Theory of operation:
+ * Each search creates a {@link CCSUsage.Builder}, which can be updated during the progress of the search request,
+ * and then it instantiates a {@link CCSUsage} object when the request is finished.
+ * That object is passed to {@link #updateUsage(CCSUsage)} on the request processing end (whether successful or not).
+ * The {@link #updateUsage(CCSUsage)} method will then update the internal counters and metrics.
+ * <br>
+ * When we need to return the current state of the telemetry, we can call {@link #getCCSTelemetrySnapshot()} which produces
+ * a snapshot of the current state of the telemetry as {@link CCSTelemetrySnapshot}. These snapshots are additive so
+ * when collecting the snapshots from multiple nodes, an empty snapshot is created and then all the node's snapshots are added
+ * to it to obtain the summary telemetry.
+ */
+public class CCSUsageTelemetry {
+
+    /**
+     * Result of the request execution.
+     * Either "success" or a failure reason.
+     */
+    public enum Result {
+        SUCCESS("success"),
+        REMOTES_UNAVAILABLE("remotes_unavailable"),
+        CANCELED("canceled"),
+        NOT_FOUND("not_found"),
+        TIMEOUT("timeout"),
+        CORRUPTION("corruption"),
+        SECURITY("security"),
+        // May be helpful if there's a lot of other reasons, and it may be hard to calculate the unknowns for some clients.
+        UNKNOWN("other");
```

Comment:
is there a relation between these values and status code? Could we somehow reuse existing status codes instead of introducing new concepts?

Reply:
Do you mean HTTP status codes? Then there's no relationship. Or some other status codes? Then I don't know - please point me to where we have some and I'll see if they could be unified.

Reply:
To clarify the intent here - these codes are meant to represent various search failure reasons classified into broad groups for analysis by e.g. PM. I couldn't find any pre-existing codes that match this use case but certainly would be glad to re-use if such ones exist.

Reply:
I did mean http status codes, that many times are carried around as part of our own exceptions (`status` method commonly). I see that there isn't a 1:1 mapping between this new enum and the existing exceptions. That's ok.